src/CMakeLists.txt

```diff
@@ -672,6 +673,7 @@ add_executable(ceph-mon ${ceph_mon_srcs}
   $<TARGET_OBJECTS:common_texttable_obj>)
 add_dependencies(ceph-mon erasure_code_plugins)
 target_link_libraries(ceph-mon mon os global-static ceph-common
+  legacy-option-headers
```

Comment:
@cbodley hi Casey, thank you for your review and approval!

"no" in this context, what we want to address here is the build dependency of a compilation units included by a certain target. in this case, `ceph-mon` is compiled from `ceph_mon.cc`, and then linked against some other libraries, they in turn link against common. but the generator (ninja or make) could build `ceph_mon.cc` even before these libraries are compiled and linked. in that case, if the headers are not generated, we have FTBFS. so the one of the formal ways to find the dependencies is to precompile the source files and to find the dependencies from cmake targets to `legacy_config_opts.h` by referencing the compiling database (`compile_commands.json`).

Reply:
> but the generator (ninja or make) could build `ceph_mon.cc` even before these libraries are compiled and linked

that's surprising, especially when [GENERATED](https://cmake.org/cmake/help/latest/prop_sf/GENERATED.html) headers are part of that dependency

Reply:
> but the generator (ninja or make) could build `ceph_mon.cc` even before these libraries are compiled and linked

i don't think this is even true in general. from [add_dependencies()](https://cmake.org/cmake/help/latest/command/add_dependencies.html):

> Makes a top-level \<target\> depend on other top-level targets to ensure that they build before \<target\> does.

Reply:
@cbodley i agree that if a target depends on some the GENERATED headers, the headers are generated before building the target. but could you please point me to the place where we define the dependency from `ceph-mon` to the headers? the linkages happens at link time, unless the linked library is an interface library -- that's exactly what i am doing in this PR.

Reply:
ceph-mon -> ceph-common -> legacy-option-headers

Reply:
i think that's a link time dependency. 

```cmake
target_link_libraries(ceph-common ${ceph_common_deps})
```

`ceph-common` is an archive of a bunch of .o files. compiler does not have to build this .a before compiling `ceph_mon.cc` into `ceph_mon.cc.o`. this is not the kind of dependency that you explained above.

Reply:
does cmake have a distinction for 'link-time dependency'? target_link_libraries() isn't just for linkage, it's also for headers

Reply:
yeah, i think that's for the interface library. or library which exposes header files with interface properties (using something like `target_sources(... INTERFACE ...)`. again, if the dependency chain from executable to .a to another .a and all the way down to .h had worked, we would not need to worry about the problem, and this PR should be closed, ... even the first 2 commits in it would not be necessary.

but i doubt it.

Reply:
oh, i see. so you are referencing:

```cmake
target_link_libraries(ceph-common legacy-option-headers)
```

but i was staring at https://github.com/ceph/ceph/blob/720f8e954cd9c4d6163a1fee379f1c286b4f5c7e/src/ceph_mon.cc#L22

so you are correct.

Reply:
thank you for your insights. i reverted a couple changes in the last commit.

Reply:
thanks Kefu! still has my approval 👍